File: examples/c-sharp-examples/GlobalRegistration.dox.cs
using System.Reflection;

public class GlobalRegistrationExample
{
    static void PrintStats(MR.MultiwayICP icp)
    {
        ulong numActivePairs = icp.getNumActivePairs();
        Console.WriteLine($"Number of samples: {icp.getNumSamples()}");
        Console.WriteLine($"Number of active pairs: {numActivePairs}");

        if (numActivePairs > 0)
        {
            double p2ptMetric = icp.getMeanSqDistToPoint();
            double p2ptInaccuracy = icp.getMeanSqDistToPoint(p2ptMetric);
            Console.WriteLine($"RMS point-to-point distance: {p2ptMetric} ± {p2ptInaccuracy}");

            double p2plMetric = icp.getMeanSqDistToPlane();
            double p2plInaccuracy = icp.getMeanSqDistToPlane(p2ptMetric);
            Console.WriteLine($"RMS point-to-plane distance: {p2plMetric} ± {p2plInaccuracy}");
        }
    }
    public static void Run(string[] args)
    {
        if (args.Length < 4)
        {
            Console.WriteLine("Usage: {0} GlobalRegistrationExample INPUT1 INPUT2 [INPUTS...] OUTPUT", Assembly.GetExecutingAssembly().GetName().Name);
            return;
        }

        try
        {
            int inputNum = args.Length - 2;
            MR.Vector_MRMeshOrPointsXf_MRObjId inputs = new();
            MR.Box3f maxBBox = new();
            for (int i = 0; i < inputNum; ++i)
            {
                var pc = MR.PointsLoad.fromAnySupportedFormat(args[i + 1]);
                System.Runtime.InteropServices.GCHandle.Alloc(pc);
                MR.MeshOrPointsXf obj = new MR.MeshOrPointsXf(pc, new MR.AffineXf3f());
                inputs.pushBack(obj);
                maxBBox.include(obj.obj.computeBoundingBox());
            }

            MR.MultiwayICPSamplingParameters samplingParams = new();
            samplingParams.samplingVoxelSize = maxBBox.diagonal() * 0.03f;

            MR.MultiwayICP icp = new(inputs, samplingParams);
            MR.ICPProperties iCPProperties = new();
            icp.setParams(iCPProperties);
            icp.updateAllPointPairs();
            PrintStats(icp);

            Console.WriteLine("Calculating transformations...");
            MR.Vector_MRAffineXf3f_MRObjId xfs = icp.calculateTransformations();
            PrintStats(icp);
            MR.PointCloud output = new();
            for (int i = 0; i < inputNum; ++i)
            {
                MR.ObjId id = new(i);
                var xf = xfs[id];
                for (ulong j = 0; j < inputs[id].obj.points().size(); j++)
                    output.addPoint(xf.call(inputs[id].obj.points()[new MR.VertId(j)]));
            }

            MR.PointsSave.toAnySupportedFormat(output, args[args.Length - 1]);
        }
        catch (Exception e)
        {
            Console.WriteLine("Error: {0}", e.Message);
        }
    }
}

File: examples/js-examples/GlobalRegistration.dox.js
import createMeshLib from '@meshinspector/meshlib';

const ml = await createMeshLib();

// the global registration can be applied to meshes and point clouds;
// to keep the sample simple, we work with point clouds only
const inputPaths = ['cloud1.ply', 'cloud2.ply', 'cloud3.ply'];
const outputPath = 'registered.ply';

// a DisposableStack frees the per-input WASM handles (there is a variable number of them) at the
// end of the scope, in reverse order, the same way `using` frees a single handle
using stack = new DisposableStack();

// as ICP and MultiwayICP accept both meshes and point clouds, each input is wrapped in a
// MeshOrPointsXf; the wrapper references the source cloud, so the clouds must outlive it (the
// stack, declared first, is disposed last)
using identityXf = new ml.AffineXf3f();
const inputs = [];
const objects = [];
let maxVolume = -Infinity;
let baseDiagonal = 0;
for (const path of inputPaths) {
  const cloud = stack.use(ml.PointsLoad.fromAnySupportedFormat(path));
  inputs.push(cloud);

  using bbox = cloud.computeBoundingBox();
  const volume = bbox.volume();
  if (volume > maxVolume) {
    maxVolume = volume;
    baseDiagonal = bbox.diagonal();
  }

  const mop = stack.use(ml.MeshOrPoints.fromPoints(cloud));
  objects.push(stack.use(new ml.MeshOrPointsXf(mop, identityXf)));
}

// set the sampling voxel size relative to the largest input
using samplingParams = new ml.MultiwayICPSamplingParameters();
samplingParams.samplingVoxelSize = baseDiagonal * 0.03;

using icp = new ml.MultiwayICP(objects, samplingParams);
using icpProps = new ml.ICPProperties();
icp.setParams(icpProps);

// gather statistics
icp.updateAllPointPairs();
console.log(`Samples: ${icp.getNumSamples()}`);
console.log(`Active point pairs: ${icp.getNumActivePairs()}`);

// one transformation per input object
const xfs = icp.calculateTransformations();

// apply each transformation to its cloud and merge everything into one output cloud
using output = new ml.PointCloud();
for (let i = 0; i < inputs.length; ++i) {
  using xf = xfs[i];
  using srcPoints = inputs[i].points;
  const coords = srcPoints.toArray();
  for (let p = 0; p < coords.length; p += 3)
    output.addPoint(xf.apply({ x: coords[p], y: coords[p + 1], z: coords[p + 2] }));
}

ml.PointsSave.toAnySupportedFormat(output, outputPath);
